### PR TITLE
Explain merge errors in designer-friendly terms

### PR DESCRIPTION
**Pull request by simoncozens**
Currently error messages from `varLib.merge` are... probably meaningful if you're behdad, but not otherwise. This patch makes `varLib.merge` explain itself better when things go wrong.

Without the patch:

```
fontTools.varLib.errors.VarLibMergeError: ((2, [2, 1, 2, 2]), 'int', '.FeatureCount', 'FeatureList', '.FeatureList', 'GPOS', '.table', 'table_G_P_O_S_')
```

With the patch:

```
fontTools.varLib.errors.VarLibMergeError:

Couldn't merge the fonts, because some values were different, but should have
been the same. This happened while performing the following operation:
.table.FeatureList.FeatureCount

The problem is likely to be in Simple Two Axis Bold:
Expected to see .FeatureCount==2, instead saw 1
```

### CODE REVIEW

**Comment by behdad** [comment]
LGTM. Thanks!

File: Lib/fontTools/varLib/errors.py [hunk 1]
```diff
@@ -1,3 +1,22 @@
+from enum import Enum, auto
```
Comment (anthrotype) [inline Lib/fontTools/varLib/errors.py:1]: nit: you imported `enum.auto` but it seems you haven't used it

**Review by anthrotype — COMMENTED** [review]
Thanks Simon, this is very helpful! 

Some tests would have been nice too :)

**Comment by RosaWagner** [comment]
Using fontmake through builder on a font which was presenting this issue last week, now output error message is:

```
Traceback (most recent call last):
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/font_project.py", line 938, in run_from_designspace
    **kwargs,
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/font_project.py", line 1014, in _run_from_designspace_interpolatable
    designspace, output_path=output_path, output_dir=output_dir, **kwargs
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/font_project.py", line 333, in build_variable_font
    inplace=True,
  File "/Users/rosalie/env/lib/python3.7/site-packages/ufo2ft/__init__.py", line 616, in compileVariableTTF
    ttfDesignSpace, exclude=excludeVariationTables, optimize=optimizeGvar
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/__init__.py", line 926, in build
    _merge_OTL(vf, model, master_fonts, axisTags)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/__init__.py", line 619, in _merge_OTL
    merger.mergeTables(font, master_fonts, ['GSUB', 'GDEF', 'GPOS'])
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 127, in mergeTables
    for m in master_ttfs])
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 1073, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 112, in mergeThings
    self.mergeObjects(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 84, in mergeObjects
    mergerFunc(self, value, values)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 1073, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 112, in mergeThings
    self.mergeObjects(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 84, in mergeObjects
    mergerFunc(self, value, values)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 1073, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 112, in mergeThings
    self.mergeObjects(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 84, in mergeObjects
    mergerFunc(self, value, values)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 1073, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 120, in mergeThings
    "got": lst},))
fontTools.varLib.errors.VarLibMergeError: <exception str() failed>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/rosalie/env/bin/gftools-builder.py", line 83, in <module>
    builder.build()
  File "/Users/rosalie/env/lib/python3.7/site-packages/gftools/builder/__init__.py", line 145, in build
    self.build_variable()
  File "/Users/rosalie/env/lib/python3.7/site-packages/gftools/builder/__init__.py", line 246, in build_variable
    output_files = self.run_fontmake(source, args)
  File "/Users/rosalie/env/lib/python3.7/site-packages/gftools/builder/__init__.py", line 265, in run_fontmake
    FontProject().run_from_glyphs(source, **args)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/font_project.py", line 691, in run_from_glyphs
    self.run_from_designspace(designspace_path, **kwargs)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/font_project.py", line 949, in run_from_designspace
    ) from e
fontmake.errors.FontmakeError: <exception str() failed>
```

If I run only fontmake:

```
Traceback (most recent call last):
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/font_project.py", line 938, in run_from_designspace
    **kwargs,
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/font_project.py", line 1014, in _run_from_designspace_interpolatable
    designspace, output_path=output_path, output_dir=output_dir, **kwargs
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/font_project.py", line 333, in build_variable_font
    inplace=True,
  File "/Users/rosalie/env/lib/python3.7/site-packages/ufo2ft/__init__.py", line 616, in compileVariableTTF
    ttfDesignSpace, exclude=excludeVariationTables, optimize=optimizeGvar
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/__init__.py", line 926, in build
    _merge_OTL(vf, model, master_fonts, axisTags)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/__init__.py", line 619, in _merge_OTL
    merger.mergeTables(font, master_fonts, ['GSUB', 'GDEF', 'GPOS'])
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 127, in mergeTables
    for m in master_ttfs])
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 1073, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 112, in mergeThings
    self.mergeObjects(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 84, in mergeObjects
    mergerFunc(self, value, values)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 1073, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 112, in mergeThings
    self.mergeObjects(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 84, in mergeObjects
    mergerFunc(self, value, values)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 1073, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 112, in mergeThings
    self.mergeObjects(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 84, in mergeObjects
    mergerFunc(self, value, values)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 1073, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/merger.py", line 120, in mergeThings
    "got": lst},))
fontTools.varLib.errors.VarLibMergeError: <exception str() failed>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/__main__.py", line 445, in main
    project.run_from_glyphs(glyphs_path, **args)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/font_project.py", line 691, in run_from_glyphs
    self.run_from_designspace(designspace_path, **kwargs)
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/font_project.py", line 949, in run_from_designspace
    ) from e
fontmake.errors.FontmakeError: <exception str() failed>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/ttLib/ttFont.py", line 372, in __getitem__
    return self.tables[tag]
KeyError: 'name'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/rosalie/env/bin/fontmake", line 8, in <module>
    sys.exit(main())
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/__main__.py", line 486, in main
    sys.exit(f"fontmake: Error: {str(e)}")
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontmake/errors.py", line 33, in __str__
    cause = str(self.__cause__) if self.__cause__ is not None else None
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/errors.py", line 39, in __str__
    for ttf in self.merger.ttfs
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/varLib/errors.py", line 39, in <listcomp>
    for ttf in self.merger.ttfs
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/ttLib/ttFont.py", line 381, in __getitem__
    data = self.reader[tag]
  File "/Users/rosalie/env/lib/python3.7/site-packages/fontTools/ttLib/sfnt.py", line 105, in __getitem__
    entry = self.tables[Tag(tag)]
KeyError: 'name'
```

Same if I reinstall fontmake and https://github.com/fonttools/fonttools in a new venv

**Comment by anthrotype** [comment]
thanks @RosaWagner for reporting this. Do you mind filing a separate issue? This PR already got merged so the bug would need to be fixed in a new PR before the next release.

**Comment by RosaWagner** [comment]
np, done in #2224 

Didn't know if it was a real issue or if I just don't know how to pip install stuff correctly ;)

**Comment by simoncozens** [comment]
Test would have been nice but I have not found it easy to replicate these merge errors... Hopefully I will be able to derive a test case.

**Comment by alexeiva** [comment]
@simoncozen I am also testing your fonttools fork. 

As an enhancement suggestion, it would be useful to know the names of parent glyphs in question when fontmake references errors in bracket layers. ( I fixed my error by renaming the bracket layers to {205,0.5} from {210,0.5} )


```
ERROR:root:In 'XXXSans.glyphs' -> 'master_ufo/XXXSans.designspace': Generating fonts from Designspace failed: Source or instance 'XXX Black Proportional {210,0.39}' has out-of-range location for axis 'Weight': is mapped to 905.0 but must be in mapped range [300.0..900.0] (NOTE: all values are in user-space).
Traceback (most recent call last):
  File "/Users/alexeiva/virtualenvs/ftools-simon/lib/python3.9/site-packages/fontmake/font_project.py", line 960, in run_from_designspace
    self._run_from_designspace_interpolatable(
  File "/Users/alexeiva/virtualenvs/ftools-simon/lib/python3.9/site-packages/fontmake/font_project.py", line 1040, in _run_from_designspace_interpolatable
    self.build_variable_font(
  File "/Users/alexeiva/virtualenvs/ftools-simon/lib/python3.9/site-packages/fontmake/font_project.py", line 331, in build_variable_font
    font = ufo2ft.compileVariableTTF(
  File "/Users/alexeiva/virtualenvs/ftools-simon/lib/python3.9/site-packages/ufo2ft/__init__.py", line 634, in compileVariableTTF
    varfont = varLib.build(
  File "/Users/alexeiva/src/fonttools/Lib/fontTools/varLib/__init__.py", line 881, in build
    ds = load_designspace(designspace)
  File "/Users/alexeiva/src/fonttools/Lib/fontTools/varLib/__init__.py", line 779, in load_designspace
    raise VarLibValidationError(
fontTools.varLib.errors.VarLibValidationError: Source or instance 'XXX Black Proportional {210,0.39}' has out-of-range location for axis 'Weight': is mapped to 905.0 but must be in mapped range [300.0..900.0] (NOTE: all values are in user-space).

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/alexeiva/virtualenvs/ftools-simon/lib/python3.9/site-packages/fontmake/__main__.py", line 482, in main
    project.run_from_glyphs(glyphs_path, **args)
  File "/Users/alexeiva/virtualenvs/ftools-simon/lib/python3.9/site-packages/fontmake/font_project.py", line 711, in run_from_glyphs
    self.run_from_designspace(designspace_path, **kwargs)
  File "/Users/alexeiva/virtualenvs/ftools-simon/lib/python3.9/site-packages/fontmake/font_project.py", line 974, in run_from_designspace
    raise FontmakeError(
fontmake.errors.FontmakeError: In 'XXXSans.glyphs' -> 'master_ufo/XXXSans.designspace': Generating fonts from Designspace failed: Source or instance 'XXX Black Proportional {210,0.39}' has out-of-range location for axis 'Weight': is mapped to 905.0 but must be in mapped range [300.0..900.0] (NOTE: all values are in user-space).
```


--verbose DEBUG output:

```
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-LightProportional-{56,0.39}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-LightProportional-{56,0.4}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-LightProportional-{56,0.49}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-LightProportional-{56,0.5}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-LightProportional-{56,0.51}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-LightProportional-{56,0.59}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-LightProportional-{56,0.6}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BoldProportional
DEBUG:ufo2ft.filters.base:Took 0.012s to run DecomposeComponentsFilter on 92 glyphs
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BoldProportional-{137,0.39}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BoldProportional-{137,0.4}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BoldProportional-{137,0.59}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BoldProportional-{137,0.6}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BlackProportional
DEBUG:ufo2ft.filters.base:Took 0.011s to run DecomposeComponentsFilter on 92 glyphs
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BlackProportional-{210,0.39}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BlackProportional-{210,0.4}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BlackProportional-{210,0.49}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BlackProportional-{210,0.5}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BlackProportional-{210,0.51}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BlackProportional-{210,0.59}
INFO:ufo2ft.filters.base:Running DecomposeComponentsFilter on XXXSans-BlackProportional-{210,0.6}
```

**Comment by simoncozens** [comment]
What's interesting to me is that the error happened when loading the designspace - that UFO should never have been produced! In fact, glyphsLib has a check for *something like* this:

```
            if not bracket_axis_min <= bracket_crossover <= bracket_axis_max:
                raise ValueError(
                    "Glyph {glyph_name}: Bracket layer {layer_name} must be within the "
                    "design space bounds of the {bracket_axis_name} axis: minimum "
                    "{bracket_axis_minimum}, maximum {bracket_axis_maximum}.".format(
                        glyph_name=glyph_name,
                        layer_name=layer.name,
                        bracket_axis_name=bracket_axis.name,
                        bracket_axis_minimum=bracket_axis_min,
                        bracket_axis_maximum=bracket_axis_max,
                    )
                )
```

Maybe we should not just check the designspace bounds but also the mapped userspace coordinate too? (I might sound like I know what I'm talking about, but I don't really understand bracket layers and anything that involves coordinate mapping confuses the heck out of me.) At any rate I see this as a glyphsLib bug, if it's producing out-of-range layers.

One for @madig, I think.

**Comment by madig** [comment]
@alexeiva I'd need to have a look at the .glyphs file to judge if it's a bug or not, but the error message contains:

> ERROR:root:In 'XXXSans.glyphs' -> 'master_ufo/XXXSans.designspace': Generating fonts from Designspace failed: Source or instance 'XXX Black Proportional {210,0.39}' has out-of-range location for axis 'Weight': is mapped to 905.0 but must be in mapped range [300.0..900.0] (NOTE: all values are in user-space).

Meaning you had your heaviest weight at 205 design == 900 user coordinate. Glyphs can extrapolate a bit, fontmake cannot. Your brace layers therefore have to be kept inside the axis limits.

**Comment by simoncozens** [comment]
I think it is a bug, because glyphsLib produced a file that varLib says is invalid. The problem should have been caught earlier; if it was caught in glyphsLib, you would be able to report the glyph that caused the problem.

**Comment by alexeiva** [comment]
@madig Initially I had 210 == 900, then updated to 205 == 900, but forgot to update the bracket layers. 

Here is an opensource font where I reproed the same issue:
[Arima-bracket-test.glyphs.zip](https://github.com/fonttools/fonttools/files/6204391/Arima-bracket-test.glyphs.zip)

**Pull request merged**